### PR TITLE
Implement Copy for matrices and vetors up to dimension 6 only.

### PR DESCRIPTION
**Pull request by sebcrozet**
See the discussion in #334.

I'm not going to merge this before a week because I which to let anyone strongly opposed to this change to provide a feedback.

### CODE REVIEW

**Comment by jswrenn** [comment]
As I described [here](https://github.com/sebcrozet/nalgebra/issues/334#issuecomment-386647661), I oppose this change on the basis of conforming to the documented scope of `Copy`.

As an alternative: We should keep only the `&Matrix` × `&Matrix` binop implementation, and remove the `Matrix` × `Matrix`, `Matrix` × `&Matrix` and `&Matrix` × `Matrix` implementations. These other implementations just call out to the `&Matrix` × `&Matrix` implementation and we _assumed_ that the copy would therefore be optimized away. [It is **not** optimized away](https://github.com/sebcrozet/nalgebra/issues/334#issuecomment-386668587), and thus these other implementations are silent performance traps.

Removing these extraneous binop implementations would not significantly degrade ergonomics. For one, the problem can be resolved by users by adding a single character (`&`). Second, the error message users would receive if we took this approach would be much more helpful than the error message they would receive it we removed `Copy`. When these binop implementations are removed, rustc produces a very helpful error message to users who inadvertently pass `Matrix` rather than `&Matrix` as an operand to a binop that recommends that they pass a reference instead. For most users, _this_ is the right thing to do, _not_ add a `.clone()`.

In short this approach:
 - better corresponds with how the Rust documentation says `Copy` should be used
 - presents more helpful error messages to users

**Comment by sebcrozet** [comment]
Closing this as it was not a good idea.

**Comment by Nufflee** [comment]
> Closing this beacause it was not a good idea.

@sebcrozet excuse my ignorance but do you mind expanding on why exactly this would be a bad idea? Assuming that the compiler bug is fixed or all binops are changed to take `&Matrix`, why not implement `Copy` for small matrices whose element type is also `Copy`? I feel like doing so would improve the ergonomics of `nalgebra-glm` (I only really have experience with it) as most functions wouldn't have to take `Vec{N}`s by borrow any more.

**Comment by Nufflee** [comment]
@Andlon: ah, I see, thanks. It remains a mystery to me why functions such as [nalgebra_glm::vec2_to_vec3](https://docs.rs/nalgebra-glm/latest/nalgebra_glm/fn.vec2_to_vec3.html) take arguments by borrow then.

**Comment by sebcrozet** [comment]
@Nufflee It’s mostly for being coherent with the the original `nalgebra` API. However, I agree that, as far as `nalgebra-glm` is conserned, we should keep it simpler and not take any references.
